File: sdk/dart/lib/src/ceph_s3.dart
import 'http_client.dart';

/// Ceph S3: S3 connection, credentials, usage, buckets, sub-users, key rotation
class CephS3Resource {
  final HttpClient httpClient;
  CephS3Resource(this.httpClient);

  /// GET /api/service/{id}/s3 - Get S3 connection info
  Future<Map<String, dynamic>> getConnectionInfo([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/s3', params);
  }

  /// GET /api/service/{id}/s3/credentials - Get S3 credentials with secret key
  Future<Map<String, dynamic>> getCredentials([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/s3/credentials', params);
  }

  /// GET /api/service/{id}/s3/usage - Get S3 usage
  Future<Map<String, dynamic>> getUsage([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/s3/usage', params);
  }

  /// GET /api/service/{id}/s3/metrics - Get S3 metrics
  Future<Map<String, dynamic>> getMetrics([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/s3/metrics', params);
  }

  /// GET /api/service/{id}/s3/buckets - List buckets
  Future<Map<String, dynamic>> listBuckets([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/s3/buckets', params);
  }

  /// POST /api/service/{id}/s3/buckets - Create bucket
  Future<Map<String, dynamic>> createBucket([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/s3/buckets', params);
  }

  /// DELETE /api/service/{id}/s3/buckets/{bucket} - Delete bucket
  Future<Map<String, dynamic>> deleteBucket([Map<String, dynamic>? params]) async {
    return httpClient.delete('/api/service/{id}/s3/buckets/{bucket}', params);
  }

  /// GET /api/service/{id}/s3/subusers - List sub-users
  Future<Map<String, dynamic>> listSubusers([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/s3/subusers', params);
  }

  /// POST /api/service/{id}/s3/subusers - Create sub-user
  Future<Map<String, dynamic>> createSubuser([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/s3/subusers', params);
  }

  /// DELETE /api/service/{id}/s3/subusers/{subuser} - Delete sub-user
  Future<Map<String, dynamic>> deleteSubuser([Map<String, dynamic>? params]) async {
    return httpClient.delete('/api/service/{id}/s3/subusers/{subuser}', params);
  }

  /// POST /api/service/{id}/s3/key - Rotate secret key
  Future<Map<String, dynamic>> rotateSecretKey([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/s3/key', params);
  }

}

File: sdk/julia/src/ceph_s3.jl
# Ceph S3: S3 connection, credentials, usage, buckets, sub-users, key rotation
struct CephS3Resource
    http::HttpClient
end

# GET /api/service/{id}/s3 - Get S3 connection info
function get_connection_info(r::CephS3Resource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/s3"; params = params)
end

# GET /api/service/{id}/s3/credentials - Get S3 credentials with secret key
function get_credentials(r::CephS3Resource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/s3/credentials"; params = params)
end

# GET /api/service/{id}/s3/usage - Get S3 usage
function get_usage(r::CephS3Resource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/s3/usage"; params = params)
end

# GET /api/service/{id}/s3/metrics - Get S3 metrics
function get_metrics(r::CephS3Resource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/s3/metrics"; params = params)
end

# GET /api/service/{id}/s3/buckets - List buckets
function list_buckets(r::CephS3Resource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/s3/buckets"; params = params)
end

# POST /api/service/{id}/s3/buckets - Create bucket
function create_bucket(r::CephS3Resource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/s3/buckets"; params = params)
end

# DELETE /api/service/{id}/s3/buckets/{bucket} - Delete bucket
function delete_bucket(r::CephS3Resource; kwargs...)
    params = Dict(kwargs)
    http_delete(r.http, "/api/service/{id}/s3/buckets/{bucket}"; params = params)
end

# GET /api/service/{id}/s3/subusers - List sub-users
function list_subusers(r::CephS3Resource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/s3/subusers"; params = params)
end

# POST /api/service/{id}/s3/subusers - Create sub-user
function create_subuser(r::CephS3Resource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/s3/subusers"; params = params)
end

# DELETE /api/service/{id}/s3/subusers/{subuser} - Delete sub-user
function delete_subuser(r::CephS3Resource; kwargs...)
    params = Dict(kwargs)
    http_delete(r.http, "/api/service/{id}/s3/subusers/{subuser}"; params = params)
end

# POST /api/service/{id}/s3/key - Rotate secret key
function rotate_secret_key(r::CephS3Resource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/s3/key"; params = params)
end
